### PR TITLE
Election: avoid space question

### PR DESCRIPTION
**Pull request by MeKHell**
This PR solves #1428 . 
It prevents to have: 
- blank question name
- blank election name
- blank ballot answers

We achieve this by trimming every user input and preventing users to enter whitespace-only strings or strings starting with whitespace.

This does not impact the UI. So no pictures.

### CODE REVIEW

**Comment by dayan9265** [comment]
It looks good to me, I just have a small question, what was the reason to only trim the beginning of the user input when shown and not also the end?

**Comment by dayan9265** [comment]
> > It looks good to me, I just have a small question, what was the reason to only trim the beginning of the user input when shown and not also the end?
> 
> If we also trim the end on the reacting inputs, the user won't be able to type space while they write and would need to insert the space (after they have written what they want) in the middle of their text.
> 
> But when the written text is ready to be processed (when the organizer confirms the election), we usually trim both going further.

Oh okay it makes sense, clearly it would not be very practical if the user couldn't type white space at the end of the text ^^

**Review by dayan9265 — DISMISSED** [review]
It looks good to me :)

File: fe1-web/src/features/evoting/screens/CreateElection.tsx [hunk 1]
```diff
@@ -282,7 +284,7 @@ const CreateElection = () => {
                   id === idx
                     ? {
                         ...item,
-                        ballot_options: ballot_options,
+                        ballot_options: ballot_options.filter((option) => option.trim() !== ''),
```
Comment (Tyratox) [inline fe1-web/src/features/evoting/screens/CreateElection.tsx:287]: Why omit empty ballot options here? I understand that we don't want empty ballot options in the final object but showing an error to the user is a better option than just omitting data without telling the user.

It works because the TextInputList is not a controlled component (https://stackoverflow.com/a/42522792/2897827) but a future refactor should probably change this and ensure the state is in sync with what is displayed by the `TextInputList` component.

File: fe1-web/src/features/evoting/screens/CreateElection.tsx [hunk 2]
```diff
@@ -302,7 +304,7 @@ const CreateElection = () => {
           {STRINGS.event_creation_must_be_connected}
         </Text>
       )}
-      {electionName === '' && (
+      {electionName.trim() === '' && (
```
Comment (Tyratox) [inline fe1-web/src/features/evoting/screens/CreateElection.tsx:307]: Yes, this is great! If something is off usually it is best to tell the user and let them fix it. Trimming whitespace is fine without prompting the user is fine as it does not change the meaning but having an empty election name is something they have to fix.

File: fe1-web/src/features/evoting/screens/CreateElection.tsx [hunk 3]
```diff
@@ -74,7 +74,9 @@ const isQuestionInvalid = (question: NewQuestion): boolean =>
  * @param questions The list of questions
  */
 const haveQuestionsSameTitle = (questions: NewQuestion[]): boolean => {
-  const questionTitles = questions.map((q: NewQuestion) => q.question);
+  const questionTitles = questions
+    .map((q: NewQuestion) => q.question.trim())
```
Comment (Tyratox) [inline fe1-web/src/features/evoting/screens/CreateElection.tsx:78]: This is great, in the final questions will be trimmed and thus we should check for equality where they are trimmed. But since we now make a transformation between the data displayed to the user and what will be used to create the election we should ensure we perform the checks on the same data that will then be used to create the election. Otherwise the two things might differ and we have code duplication.

One way to solve is is by separating the values displayed to the user and the ones used for processing later. For instance you can create an additional `electionName` and `questions` variables and rename the old variables to something like `electionNameInput` and `questionsInput`.

E.g.
```typescript
const [questionsInput, setQuestionsInput] = useState<NewQuestion[]>([EMPTY_QUESTION]);
...


const questions = useMemo(() => {
  return questionsInput.map((q: NewQuestion) => {...q, question: q.question.trim()});
}, [questionsInput]);
```

In this case `questionsInput` represents what is entered by the user and `questions` what we use for processing. All checks should be performed on `questions` and tell the user what they need to change in `questionsInput` to make the input valid. `useMemo` updates its value if one of it's dependencies (in the array) change.

This way we're sure that the data used for processing (`questions`, derived from `questionsInput` ) is valid.

Does this make sense?

File: fe1-web/src/features/evoting/screens/CreateElection.tsx [hunk 4]
```diff
@@ -74,7 +74,9 @@ const isQuestionInvalid = (question: NewQuestion): boolean =>
...
+    .filter((string) => string !== '');
```
Comment (Tyratox) [inline fe1-web/src/features/evoting/screens/CreateElection.tsx:79]: Again, this should be done by the user as this is not something one would expect to happen automatically.

File: fe1-web/src/features/evoting/screens/CreateElection.tsx [hunk 5]
```diff
@@ -151,7 +153,7 @@ const CreateElection = () => {
   // Confirm button only clickable when the Name, Question and 2 Ballot options have values
   const confirmButtonEnabled: boolean =
     isConnected === true &&
-    electionName !== '' &&
+    electionName.trim() !== '' &&
```
Comment (Tyratox) [inline fe1-web/src/features/evoting/screens/CreateElection.tsx:156]: Same comment as above. Try to separate the value displayed by the user and the one used for processing and perform the check on the one used for processing.

At the moment you perform the processing and the checks in the same place.

File: fe1-web/src/features/evoting/screens/CreateElection.tsx [hunk 6]
```diff
@@ -105,9 +107,9 @@ const createElection = (
   // compute the id for all questions and add the write_in property
   const questionsWithId = questions.map((item) =>
     Question.fromState({
-      id: Hash.fromArray(EventTags.QUESTION, electionId, item.question).toString(),
-      question: item.question,
-      ballot_options: item.ballot_options,
+      id: Hash.fromArray(EventTags.QUESTION, electionId, item.question.trim()).toString(),
+      question: item.question.trim(),
+      ballot_options: item.ballot_options.map((s) => s.trim()),
```
Comment (Tyratox) [inline fe1-web/src/features/evoting/screens/CreateElection.tsx:112]: Same comment as above. Try to create the "ready for processing" values using `useMemo`, perform the checks on these and then directly use them here.

If you perform processing here you cannot be sure whether the resulting data satisfies the desired properties.

**Review by Tyratox — CHANGES_REQUESTED** [review]
Cool that you made it work but honestly it is difficult to tell whether all places were covered and whether there are not missed edge cases.

I would try to take a step back and separate the processing of the data from the checks. You can do so by having separate variables, one for the raw user input data and one for the processed (e.g. trimmed) data. Then you can perform the checks on the processed/trimmed data and if that data satisfies the desired properties, directly use that data without performing (in the best case the same) processing again.

I made a lot of comments so it might not everywhere be clear what I mean. I am happy to do a quick zoom call tomorrow or on Tuesday if anything is unclear.

File: fe1-web/src/core/components/TextInputList.tsx [hunk 7]
```diff
@@ -20,7 +20,7 @@ const TextInputList = (props: IPropTypes) => {
   const updateParent = (options: { id: number; value: string }[]) => {
     // Gets the distinct options which are not empty ('')
     const distinctValues = [...new Set(options.map((option) => option.value))].filter(
-      (value) => value !== '',
+      (value) => value.trim() !== '',
     );
```
[the remark this answers is not in the document]
Comment (Xelowak) [reply fe1-web/src/core/components/TextInputList.tsx:24]: I agree. This component is from `core`, meaning that it should be generic to allow most uses in the app. When we need a list of text inputs, we should look into `TextInputLIst`. This kind of behaviour you are adding may not be expected at first, leading in something too specific for this kind of component.

**Review by Xelowak — COMMENTED** [review]
@Tyratox 's comments are really relevant and I honestly don't have much to add.

I feel like the trimming is done in a lot of different locations, when it could probably be more centralized. As @Tyratox suggested, I think it is better to create values that are directly trimmed and perform checks on them, rather than trimming everything before processing. This may solve the problem of trimming strings everywhere.

File: fe1-web/src/features/evoting/screens/CreateElection.tsx [hunk 8]
```diff
@@ -220,7 +222,7 @@ const CreateElection = () => {
       </Text>
       <Input
         value={electionName}
-        onChange={setElectionName}
+        onChange={(s) => setElectionName(s.trimStart())}
```
[the remark this answers is not in the document]
Comment (MeKHell) [reply fe1-web/src/features/evoting/screens/CreateElection.tsx:225]: The idea is to give a live feedback to the user so that he understand that whitespace only strings are not allowed

File: fe1-web/src/features/evoting/screens/CreateElection.tsx [hunk 9]
```diff
@@ -255,15 +257,15 @@ const CreateElection = () => {
...
             testID={`question_selector_${idx}`}
             onChange={(text: string) =>
               setQuestions((prev) =>
                 prev.map((item, id) =>
-                  id === idx
+                  id === idx && text.trimStart() !== ''
                     ? {
                         ...item,
-                        question: text,
+                        question: text.trimStart(),
```
[the remark this answers is not in the document]
Comment (MeKHell) [reply fe1-web/src/features/evoting/screens/CreateElection.tsx:268]: See https://github.com/dedis/popstellar/pull/1456#discussion_r1143079249

File: fe1-web/src/features/evoting/screens/CreateElection.tsx [hunk 10]
```diff
@@ -255,15 +257,15 @@ const CreateElection = () => {
...
           </Text>
           <Input
-            value={questions[idx].question}
+            value={value.question}
             testID={`question_selector_${idx}`}
             onChange={(text: string) =>
               setQuestions((prev) =>
                 prev.map((item, id) =>
-                  id === idx
+                  id === idx && text.trimStart() !== ''
```
[the remark this answers is not in the document]
Comment (MeKHell) [reply fe1-web/src/features/evoting/screens/CreateElection.tsx:265]: Indeed. The idea was that the `InputList` will remove it. Which might be wrong.
In any case, with the refactoring of the `InputList`, all the logic has been centralized in `CreateElection` and this part will also change.

**Comment by MeKHell** [comment]
Based on the comment above, I decided to refactor the following components: 
 - RemovableTextInput
 - TextInputList
 in order to correctly test the new CreateElectionScreen (and that the bug as disappeared), I added the ability to store default questions in the election reducer.

This is not the optimal solution. But it might be useful if we later add a feature to set default questions.

The other solution would have been to allow to pass the default questions as navigation parameters. But this would be even less logic and useful for later development.

File: fe1-web/src/features/evoting/screens/CreateElection.tsx [hunk 11]
```diff
@@ -141,22 +153,46 @@ const CreateElection = () => {
...
+
   // Confirm button only clickable when the Name, Question and 2 Ballot options have values
   const confirmButtonEnabled: boolean =
     isConnected === true &&
-    electionName !== '' &&
-    !questions.some(isQuestionInvalid) &&
-    haveQuestionsSameTitle(questions);
+    electionName.trim() !== '' &&
+    areQuestionsValid(trimmedQuestions) && // Checks there are enough valid questions to open an election
+    !questions.some(isNonEmptyQuestionInvalid) && // Checks that there are no questions that will be silently removed
```
Comment (Tyratox) [inline fe1-web/src/features/evoting/screens/CreateElection.tsx:183]: If this is the only place `isNonEmptyQuestionInvalid` is used you might even call it `isSilentlyRemoved`? Then the could would directly read like your comment :)

File: fe1-web/src/features/evoting/screens/CreateElection.tsx [hunk 12]
```diff
@@ -141,22 +153,46 @@ const CreateElection = () => {
...
+  }, [questions]);
+
   // Confirm button only clickable when the Name, Question and 2 Ballot options have values
   const confirmButtonEnabled: boolean =
     isConnected === true &&
-    electionName !== '' &&
-    !questions.some(isQuestionInvalid) &&
-    haveQuestionsSameTitle(questions);
+    electionName.trim() !== '' &&
+    areQuestionsValid(trimmedQuestions) && // Checks there are enough valid questions to open an election
```
Comment (Tyratox) [inline fe1-web/src/features/evoting/screens/CreateElection.tsx:182]: If this is the only check performed by this function, feel free to call it like that or in this particular case write it inline and keep the comment :)

File: fe1-web/src/features/evoting/screens/CreateElection.tsx [hunk 13]
```diff
@@ -141,22 +153,46 @@ const CreateElection = () => {
...
+    !questions.some(isNonEmptyQuestionInvalid) && // Checks that there are no questions that will be silently removed
+    !trimmedQuestions.some(areBallotOptionsInvalid) &&
```
Comment (Tyratox) [inline fe1-web/src/features/evoting/screens/CreateElection.tsx:184]: Picky detail but I think it could read even a bit nicer :)
```suggestion
    !trimmedQuestions.some(hasInvalidBallotOptions) &&
```

File: fe1-web/src/features/evoting/screens/CreateElection.tsx [hunk 14]
```diff
@@ -50,38 +50,49 @@ type NavigationProps = CompositeScreenProps<
...
+const areBallotOptionsInvalid = (question: NewQuestion): boolean =>
+  question.question !== '' &&
+  (new Set(question.ballot_options).size !== question.ballot_options.length ||
+    question.ballot_options.length < MIN_BALLOT_OPTIONS);
+
+/**
+ * Checks whether a question title is not unique withcreateEin a list of questions
  * @param questions The list of questions
  */
 const haveQuestionsSameTitle = (questions: NewQuestion[]): boolean => {
```
Comment (Tyratox) [inline fe1-web/src/features/evoting/screens/CreateElection.tsx:85]: You are checking whether they are unique, right?
```suggestion
const haveUniqueQuestionTitles = (questions: NewQuestion[]): boolean => {
```

File: fe1-web/src/features/evoting/screens/CreateElection.tsx [hunk 15]
```diff
@@ -141,22 +153,46 @@ const CreateElection = () => {
...
+        return {
+          ...item,
+          question: item.question.trim(),
+          ballot_options: item.ballot_options
+            .map((val) => val.trim())
+            .filter((value) => value !== ''),
+        };
+      })
+      .filter((question) => question.question !== '');
+  }, [questions]);
```
Comment (Tyratox) [inline fe1-web/src/features/evoting/screens/CreateElection.tsx:176]: 🥳👏

File: fe1-web/src/features/evoting/screens/CreateElection.tsx [hunk 16]
```diff
@@ -141,22 +153,46 @@ const CreateElection = () => {
...
+    !questions.some(isNonEmptyQuestionInvalid) && // Checks that there are no questions that will be silently removed
+    !trimmedQuestions.some(areBallotOptionsInvalid) &&
+    haveQuestionsSameTitle(trimmedQuestions);
 
   const onCreateElection = () => {
-    createElection(currentLao.id, version, electionName, questions, startTime, endTime)
+    createElection(
+      currentLao.id,
+      version,
+      electionName.trim(),
```
Comment (Tyratox) [inline fe1-web/src/features/evoting/screens/CreateElection.tsx:191]: I guess it's fine if it is just twice but it might also be nice to have a `trimmedElectionName` analogous  to `trimmedQuestions`?

File: fe1-web/src/features/evoting/screens/CreateElection.tsx [hunk 17]
```diff
@@ -220,7 +256,7 @@ const CreateElection = () => {
       </Text>
       <Input
         value={electionName}
-        onChange={setElectionName}
+        onChange={(s) => setElectionName(s)}
```
Comment (Tyratox) [inline fe1-web/src/features/evoting/screens/CreateElection.tsx:259]: Is there a reason for this change?

File: fe1-web/src/features/evoting/screens/CreateElection.tsx [hunk 18]
```diff
@@ -250,12 +286,12 @@ const CreateElection = () => {
       {questions.map((value, idx) => (
         // FIXME: Do not use index in key
         // eslint-disable-next-line react/no-array-index-key
-        <View key={idx.toString()}>
+        <View key={idx.toString()} style={{ marginBottom: Spacing.x2 }}>
```
Comment (Tyratox) [inline fe1-web/src/features/evoting/screens/CreateElection.tsx:289]: Inline styles are discouraged, I would create a small StyleSheet at the top of the file with a `questionWrapper` style :)

File: fe1-web/src/features/evoting/screens/CreateElection.tsx [hunk 19]
```diff
@@ -250,12 +286,12 @@ const CreateElection = () => {
       {questions.map((value, idx) => (
```
Comment (Tyratox) [inline fe1-web/src/features/evoting/screens/CreateElection.tsx:286]:
```suggestion
      {questions.map((question, idx) => (
```

?

File: fe1-web/src/features/evoting/screens/CreateElection.tsx [hunk 20]
```diff
@@ -269,27 +305,43 @@ const CreateElection = () => {
...
+            values={value.ballot_options}
             placeholder={STRINGS.election_create_option_placeholder}
             onChange={(ballot_options: string[]) =>
               setQuestions((prev) =>
                 prev.map((item, id) =>
                   id === idx
                     ? {
                         ...item,
-                        ballot_options: ballot_options,
+                        ballot_options: ballot_options.filter((option) => option !== ''),
```
Comment (Tyratox) [inline fe1-web/src/features/evoting/screens/CreateElection.tsx:324]: This likely causes a ballot option to disappear when it's text is deleted even if it is not the last one..? For `trimmedQuestions` you filter them a second time anyway 😄

File: fe1-web/src/features/evoting/screens/CreateElection.tsx (hunk 2 above)
Comment (Tyratox) [reply fe1-web/src/features/evoting/screens/CreateElection.tsx:357]: Okay you have to do trimming here again, I think having a `trimmedElectionName` would definitely be a good idea 😅

File: fe1-web/src/features/evoting/screens/__tests__/CreateElection.test.tsx [hunk 21]
```diff
@@ -45,10 +72,96 @@ describe('CreateElection', () => {
...
+        <FeatureContext.Provider value={contextValue}>
+          <MockNavigator component={CreateElection} />
+        </FeatureContext.Provider>
+      </Provider>,
+    );
+    const electionName = getByTestId('election_name_selector');
+    const question = getByTestId('question_selector_0');
+    fireEvent.changeText(electionName, '     ');
+    fireEvent.changeText(question, '        ');
+    expect(toJSON()).toMatchSnapshot();
```
Comment (Tyratox) [inline fe1-web/src/features/evoting/screens/__tests__/CreateElection.test.tsx:96]: I think waiting here is necessary for the snapshot to properly reflect the changes? 😅

File: fe1-web/src/features/evoting/screens/__tests__/CreateElection.test.tsx [hunk 22]
```diff
@@ -45,10 +72,96 @@ describe('CreateElection', () => {
...
+      },
+      {
+        question: '   Question     1        ',
+        ballot_options: ['    Answer 0    ', '  Answer 1  ', 'will be removed'],
+        voting_method: STRINGS.election_method_Plurality,
+        id: '',
+        write_in: false,
+      },
+    ];
+    jest.setSystemTime(Date.UTC(2023, 3, 21));
```
Comment (Tyratox) [inline fe1-web/src/features/evoting/screens/__tests__/CreateElection.test.tsx:116]: Why is this necessary? If you want to workaround specific time values you can always use https://jestjs.io/docs/expect#expectanything

**Review by Tyratox — COMMENTED** [review]
Phew that was quite the refactor, great work!! 🥳🥳👏

Very happy to see you managed to get the test working!! While I agree that the workaround is not ideal I think it is fair especially given the use case you describe 😄

I added a few minor comments (a few of them are just some additional thoughts) but it looks very very good overall! 😄

**Comment by Tyratox** [comment]
Also I just saw that sonar cloud complains about the component being to big. Maybe you can find a part that can be isolated into a separate function or component :)

File: fe1-web/src/core/components/__tests__/__snapshots__/TextInputList.test.tsx.snap [hunk 23]
```diff
@@ -0,0 +1,309 @@
...
+                "borderRadius": 8,
+                "borderWidth": 1,
+                "flex": 1,
+                "padding": 8,
+                "width": 50,
+              },
+            ]
+          }
+          testID="x_option_1_input"
+          value="   test 2"
```
[the remark this answers is not in the document]
Comment (MeKHell) [reply fe1-web/src/core/components/__tests__/__snapshots__/TextInputList.test.tsx.snap:150]: No since the input is trimmed only on submission

[the remark this answers is not in the document]
Comment (Tyratox) [reply fe1-web/src/core/components/__tests__/__snapshots__/TextInputList.test.tsx.snap:150]: Ohhh you‘re right 👍👍

File: fe1-web/src/core/components/RemovableTextInput.tsx [hunk 24]
```diff
@@ -25,33 +26,40 @@ const styles = StyleSheet.create({
...
-      <PoPTouchableOpacity
-        containerStyle={styles.icon}
-        onPress={() => onRemove(id)}
-        testID={testID ? `${testID}_remove` : undefined}>
-        <PoPIcon name="delete" color={Color.primary} size={Icon.size} />
-      </PoPTouchableOpacity>
+      {/* Added for UI coherence and alignment */}
+      {!isRemovable && (
+        <PoPTouchableOpacity containerStyle={styles.icon} onPress={() => {}}>
+          <NavigationPadding />
```
[the remark this answers is not in the document]
Comment (MeKHell) [reply fe1-web/src/core/components/RemovableTextInput.tsx:42]: I think that `ButtonPadding` is a bit more specific and closer to the usage of this component.

File: fe1-web/src/features/evoting/screens/CreateElection.tsx (hunk 19 above)
Comment (MeKHell) [reply fe1-web/src/features/evoting/screens/CreateElection.tsx:286]: This was avoided to not have some `question.question` few lines below, I think. I am not sure if it would be clearer. Changed to a longer but better name.

File: fe1-web/src/features/evoting/screens/__tests__/CreateElection.test.tsx (hunk 21 above)
Comment (MeKHell) [reply fe1-web/src/features/evoting/screens/__tests__/CreateElection.test.tsx:96]: It is not necessary.

File: fe1-web/src/features/evoting/screens/__tests__/CreateElection.test.tsx (hunk 22 above)
Comment (MeKHell) [reply fe1-web/src/features/evoting/screens/__tests__/CreateElection.test.tsx:116]: It is required, because I want the ID to be checked too. And it depends on the current time.

Comment (Tyratox) [reply fe1-web/src/features/evoting/screens/__tests__/CreateElection.test.tsx:116]: I see 👍
I guess you did not come across this yet but the global system time should already be fixed for all tests here: https://github.com/dedis/popstellar/blob/master/fe1-web/jest/setupFakeTimers.js#L6 :)

File: fe1-web/src/features/evoting/screens/CreateElection.tsx (hunk 20 above)
Comment (MeKHell) [reply fe1-web/src/features/evoting/screens/CreateElection.tsx:364]: And I just noticed that this behavior was wrong. Now I need to filter them three times because we do not have a unique question id at this stage. :sweat_smile:

File: fe1-web/src/features/evoting/screens/CreateElection.tsx [hunk 25]
```diff
@@ -63,25 +63,51 @@ const MIN_BALLOT_OPTIONS = 2;
...
+    .filter((ballot) => ballot !== '');
+  return (
+    question.ballot_options.length > 0 &&
+    (new Set(trimmedBallotOptions).size !== question.ballot_options.length ||
+      trimmedBallotOptions.length < MIN_BALLOT_OPTIONS)
+  );
+};
 
 /**
  * Checks whether a question title is not unique within a list of questions
```
Comment (Tyratox) [inline fe1-web/src/features/evoting/screens/CreateElection.tsx:99]:
```suggestion
 * Checks whether a question title is unique within a list of questions
```

File: fe1-web/src/features/evoting/screens/CreateElection.tsx [hunk 26]
```diff
@@ -125,6 +151,67 @@ const createElection = (
...
+  };
+};
+
+/**
+ * Display the error message that are not directly related to a question
+ * @param isConnected If the user isConnected to the LAO
+ * @param electionName the trimmed name of the election
+ * @param trimmedQuestions the trimmed questions of the election
+ */
+const globalErrorMessages = (
```
Comment (Tyratox) [inline fe1-web/src/features/evoting/screens/CreateElection.tsx:174]: Is there a particular reason why you used a function instead of a component?

File: fe1-web/src/features/evoting/screens/__tests__/CreateElection.test.tsx [hunk 27]
```diff
@@ -45,10 +72,97 @@ describe('CreateElection', () => {
...
+        </FeatureContext.Provider>
+      </Provider>,
+    );
+    const electionName = getByTestId('election_name_selector');
+    const question = getByTestId('question_selector_0_input');
+    const option = getByTestId('question_0_ballots_option_0_input');
+    fireEvent.changeText(electionName, '     ');
+    fireEvent.changeText(question, '        ');
+    fireEvent.changeText(option, '        ');
+    waitFor(() => expect(toJSON()).toMatchSnapshot());
```
Comment (Tyratox) [inline fe1-web/src/features/evoting/screens/__tests__/CreateElection.test.tsx:98]: Not sure this accomplished what you want 😅 Wait for executes the passed function multiple times and succeeds if the passed function succeeds once. (https://testing-library.com/docs/dom-testing-library/api-async/#waitfor)

**Review by Tyratox — APPROVED** [review]
LGTM! 😁

**Comment by Tyratox** [comment]
Just make sure the sonar cloud issues are fine and mark them accordingly :)

File: fe1-web/src/features/evoting/screens/__tests__/CreateElection.test.tsx (hunk 27 above)
Comment (MeKHell) [reply fe1-web/src/features/evoting/screens/__tests__/CreateElection.test.tsx:98]: Indeed, it is not needed, but I added it after https://github.com/dedis/popstellar/pull/1456#discussion_r1161118777, because I thought you wanted a more explicit test. My bad.

File: fe1-web/src/features/evoting/reducer/__tests__/ElectionReducer.test.ts [hunk 28]
```diff
@@ -77,20 +72,14 @@ describe('ElectionReducer', () => {
...
-            allIds: [],
-          } as ElectionReducerState,
-          updateElection(mockElectionState),
-        ),
-      ).toThrow();
-    });
+    expect(() =>
+      electionReduce(defaultElectionReducerState, updateElection(mockElectionState)),
+    ).toThrow();
+    it('throws an error when trying to update an inexistent election', () => {});
```
Comment (dayan9265) [inline fe1-web/src/features/evoting/reducer/__tests__/ElectionReducer.test.ts:82]: I suppose these lines 3 lines of test should be in the test 'throws an error when trying to update an inexistent election'?

File: fe1-web/src/features/evoting/reducer/__tests__/ElectionReducer.test.ts [hunk 29]
```diff
@@ -102,26 +91,17 @@ describe('ElectionReducer', () => {
...
-            allIds: [],
-          } as ElectionReducerState,
-          removeElection(mockElection.id),
-        ),
-      ).toThrow();
-    });
+    expect(() =>
+      electionReduce(defaultElectionReducerState, removeElection(mockElection.id)),
+    ).toThrow();
+    it('throws an error when trying to remove an inexistent election', () => {});
```
Comment (dayan9265) [inline fe1-web/src/features/evoting/reducer/__tests__/ElectionReducer.test.ts:104]: same here for the test

**Review by dayan9265 — CHANGES_REQUESTED** [review]
That was quite a big change to avoid space in question ^^, but it looks good, I have two added two comments about some tests that looked weird, but otherwise good job 👍

File: fe1-web/src/features/evoting/reducer/__tests__/ElectionReducer.test.ts (hunk 28 above)
Comment (MeKHell) [reply fe1-web/src/features/evoting/reducer/__tests__/ElectionReducer.test.ts:82]: Nice catch. This is indeed an error.

File: fe1-web/src/features/evoting/reducer/__tests__/ElectionReducer.test.ts (hunk 29 above)
Comment (MeKHell) [reply fe1-web/src/features/evoting/reducer/__tests__/ElectionReducer.test.ts]: fixed

File: fe1-web/src/features/evoting/reducer/__tests__/ElectionReducer.test.ts [hunk 30]
```diff
@@ -77,18 +72,13 @@ describe('ElectionReducer', () => {
...
     it('throws an error when trying to update an inexistent election', () => {
       expect(() =>
-        electionReduce(
-          {
-            byId: {},
-            allIds: [],
-          } as ElectionReducerState,
-          updateElection(mockElectionState),
-        ),
+        electionReduce(defaultElectionReducerState, removeElection(mockElection.id)),
```
Comment (dayan9265) [inline fe1-web/src/features/evoting/reducer/__tests__/ElectionReducer.test.ts:81]: is it wanted that the `updateElection` is now `removeElection`?

**Review by dayan9265 — APPROVED** [review]
Looks good to me, I just have a small question (see comment), but otherwise it seems perfect 😄

File: fe1-web/src/features/evoting/reducer/__tests__/ElectionReducer.test.ts (hunk 30 above)
Comment (MeKHell) [reply fe1-web/src/features/evoting/reducer/__tests__/ElectionReducer.test.ts:81]: No, I do not know how I did not catch it before. Thanks

**Review by Tyratox — APPROVED** [review]
A few small comments but I don't want to block this PR further. If you think the filtering is correct like this, feel free to handle the other comments how you prefer :)

File: fe1-web/src/features/evoting/screens/CreateElection.tsx (hunk 20 above)
Comment (MeKHell) [reply fe1-web/src/features/evoting/screens/CreateElection.tsx:417]: If we do not remove the empty ones, we will have a new empty line for every character input, since our list is now a controlled component.

Comment (Tyratox) [reply fe1-web/src/features/evoting/screens/CreateElection.tsx:417]: Ah I see you're not removing it, then it is necessary yes. I thought you changed it (https://github.com/dedis/popstellar/pull/1456/files/dc095b6cec3c47c3966620b3a5ff2a1baa70577a#r1161113914) because I am pretty sure that now there is a bug if we delete all text in the not last ballot option (it is removed). Personally I would deem this weird/bad UX but it seems it is what you want?
